### ISSUE TITLE
Short-cutting nested (model -> View Action)  (a.k.a. Components, Widgets, Gadgets)

### ISSUE THREAD

**Issue by NerdGGuy** [dmjio/miso#605]
**Is your feature request related to a problem? Please describe.**
I currently have a very large (> 100mb) single page html app I'm rendering with lucid. I use custom javascript on the browser to hide/show content as needed. Despite many of the elements being hidden the re-rending process for pages this large is very slow in browser. I'm in the process of converting this app to Miso to reduce the HTML source size by only rendering what's displayed and hopefully reduce rendering time.

**Describe the solution you'd like**
I'd like to use nested Miso (model -> View action) render functions and only render/diff when the model has changed. This may require some additional smarts in the VTree to ignore elements that haven't been re-rendered.

Say i have the following model:
```
data MyModel = MyModel SubModelA SubModelB

data SubModelA = SubModelA [String]
data SubModelB = SubModelB [Int]

```

Each multiple smaller nested Miso (model -> View Action) render functions
```
mainView :: MyModel -> View action
mainView (MyModel a b) = 
	div_ [] [
		toViewRenderer subViewA a,
		toViewRenderer subViewB b
	]

subViewA :: SubModelA -> View action
subViewA x = ...

subViewB :: SubModelB -> View action
subViewB x = ...
```

toViewRenderer can do the smarts to check if the model has changed.

**Describe alternatives you've considered**
While not perfect I also considered if the alternative was to somehow render multiple seperate Miso instances with different mounts.

**Comment by dmjio** [comment]
By default the view only gets redrawn when the model changes. The diffing algorithm ensures nothing is redrawn than hasn't been changed. This does mean that a new virtual DOM is created in memory each time the model has changed (which is relatively cheap). This does not mean a new DOM is created each time the model changes. If you can demonstrate a performance bottleneck I'd be keen to take a look.

**Comment by dmjio** [comment]
Also, I'd suggest not using `String`, but `MisoString`.

**Comment by NerdGGuy** [comment]
TL;DR - I'm starting to think this change will only provide significant benefit for me in the worst case. How often that worst case is triggered is likely to be in the 1-10% range. So in the worst case I might just turn off view rendering all together. I'll report back once I have some more real world performance 👍 . 

In my application the model is a large tree which is frequently > 100mb. The produced HTML in the worst case is 10x to 100x the size of the model. So in the worst case where the whole tree is rendered a single leaf change could cause the whole tree to be rendered when the change only effects one element. Luckily in many cases the app only renders a small section of the tree. So in the best case it's possible the model deep equality comparison on 100mb model could be a bigger impact on performance then the rendering. This is all speculative though :).

Thanks @dmjio for looking into this. Good suggestion on MisoString. I'm happy to see there's a ByteString (decodeutf8) instance which is exactly what I need :).

**Comment by dmjio** [comment]
> So in the worst case where the whole tree is rendered a single leaf change could cause the whole tree to be rendered when the change only effects one element

100MB of anything in the browser will be difficult. I'd recommend pagination, or partial rendering by fetching incrementally from the server.

**Comment by dmjio** [comment]
Please re-open if you don't deem this as resolved.
